### PR TITLE
Fix empty samplesheet

### PR DESCRIPTION
**Pull request by barrystokman**
This PR fixes production deviation #129 where demultiplexing failed due to an empty sample sheet in one of the runs directories.

**How to prepare for test**:
- [x] ssh to thalamus
- [x] install on stage:
`bash servers/resources/clinical-preproc.scilifelab.se/update-demux-stage.sh fix-empty-samplesheet`
- [x] in `checkfornewrun.bash`, uncomment line 7: `source ~/.aliases`, else the demux alias is set to the PROD version!

**How to test**:

#### TC1: fetch samplesheet for flowcell HW2C7DSXX from LIMS
- [x] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [x] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/RTAComplete.txt`
- [x] remove `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/demuxstarted.txt` if present.
- [x] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/`

**Expected test outcome**:
- [x] check that a correct samplesheet has been created in `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [x] Take a screenshot and attach or copy/paste the output.

#### TC2: samplesheet for flowcell HW2C7DSXX already exists, skip fetching from LIMS
- [x] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [x] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/RTAComplete.txt`
- [x] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/SampleSheet.csv`
- [x] `echo "I'm not empty" > /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/SampleSheet.csv`
- [x] remove `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/demuxstarted.txt` if present.
- [x] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/`

**Expected test outcome**:
- [x] The script will not fetch a new sample sheet from LIMS since a non-empy sample sheet exists.
- [x] Take a screenshot and attach or copy/paste the output.

#### TC3: samplesheet for flowcell HW2C7DSXX already exists but is EMPTY!
- [x] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [x] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/RTAComplete.txt`
- [x] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/SampleSheet.csv`
- [x] remove `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/demuxstarted.txt` if present.
- [x] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/`

#### TC4: samplesheet for flowcell not found in LIMS
- [x] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XNOFLOWCELL`
- [x] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XNOFLOWCELL/RTAComplete.txt`
- [x] remove `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XNOFLOWCELL/demuxstarted.txt` if present.
- [x] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/`
**Expected test outcome**:
- [x] The sample sheet will be empty instead of containing an error message `Samplesheet not found in LIMS!`
- [x] Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by @hassanfa 
- [x] tests executed by @barrystokman 
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by barrystokman** [comment]
#### TC1: fetch samplesheet for flowcell HW2C7DSXX from LIMS
- [ ] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [ ] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/TRAComplete.txt`
- [ ] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/ > SampleSheet.csv`

**Expected test outcome**:
- [ ] check that a correct samplesheet has been created in `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [ ] Take a screenshot and attach or copy/paste the output.

**Comment by barrystokman** [comment]
#### TC2: samplesheet for flowcell HW2C7DSXX already exists, skip fetching from LIMS
- [ ] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [ ] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/RTAComplete.txt`
- [ ] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/SampleSheet.csv`
- [ ] `echo "I'm not empty!" > /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/SampleSheet.csv`
- [ ] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/ > SampleSheet.csv`

**Expected test outcome**:
- [ ] The script will not ftech a new sample sheet from LIMS since a non-empy sample sheet exists.
- [ ] Take a screenshot and attach or copy/paste the output.

**Comment by barrystokman** [comment]
#### TC3: samplesheet for flowcell HW2C7DSXX already exists but is EMPTY!
- [ ] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX`
- [ ] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/RTAComplete.txt`
- [ ] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/SampleSheet.csv` **EMPTY SHEET!

**
- [ ] remove `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XHW2C7DSXX/demuxstarted.txt` if present.
- [ ] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/ > SampleSheet.csv`
[SampleSheet (2).zip](https://github.com/Clinical-Genomics/demultiplexing/files/4730088/SampleSheet.2.zip)

**Comment by barrystokman** [comment]
\# TODO: fix redirect in checkfornewrun.bash, but the missing samplesheet issue if fixed.

**Comment by barrystokman** [comment]
#### TC4: samplesheet for flowcell not found in LIMS
- [x] on thalamus create empty folder `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XNOFLOWCELL`
- [x] `touch /home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XNOFLOWCELL/RTAComplete.txt`
- [x] remove `/home/hiseq.clinical/STAGE/novaseq/runs/test_fix_empty_samplesheet_XNOFLOWCELL/demuxstarted.txt` if present.
- [x] run `bash /home/hiseq.clinical/STAGE/SCRIPTS/git/demultiplexing/scripts/novaseq/checkfornewrun.bash /home/hiseq.clinical/STAGE/novaseq/runs/ /home/hiseq.clinical/STAGE/novaseq/demux/`
**Expected test outcome**:
- [x] The sample sheet will be empty instead of containing an error message `Samplesheet not found in LIMS!`
- [x] Take a screenshot and attach or copy/paste the output.

image
image

**Comment by barrystokman** [comment]
Improved logging incase of empty sample sheet:
image

**Comment by barrystokman** [comment]
@moahaegglund 

This is how it works for now: 
1. Check if a sample sheet exists and if it's empty:
    - If no sample sheet exists, fetch it from LIMS
    - If a sample sheet exists and it's empty, remove it and fetch a new one from LIMS
    - If a sample sheet exists and it's not empty, do NOT fetch a new one from LIMS
1. If running `demux sheet fetch` and there's no sample sheet information in LIMS
    - send log message to script log file, including flowcell name
1. After running `demux sheet fetch`, check if the sample sheet is empty
    - if empty continue to next iteration, don't create demuxstarted.txt. Next time the script runs the demuxing for this flowcell will be attemped again, the empty sample sheet will be removed before attemting the fetch.
    - if not empty, create demuxstarted.txt, start demuxing

I think this will solve deviation #129! It would be great if you can have a look at the changes if/when you have time.

File: scripts/novaseq/checkfornewrun.bash [hunk 1]
```diff
@@ -50,17 +50,29 @@ for RUN_DIR in ${IN_DIR}/*; do
...
+            fi
 
             if [[ ! -e ${RUN_DIR}/SampleSheet.csv ]]; then
                 log "demux sheet fetch --application nova --pad --longest ${FC} > ${RUN_DIR}/SampleSheet.csv"
-                demux sheet fetch --application nova --pad --longest ${FC} > ${RUN_DIR}/SampleSheet.csv
+                demux sheet fetch --application nova --pad --longest ${FC} > ${RUN_DIR}/SampleSheet.csv 
+            fi
+
+            # exit if samplesheet is still empty after running demux sheet fetch
+            if [[ ! -s ${RUN_DIR}/SampleSheet.csv && -e ${RUN_DIR}/SampleSheet.csv ]]; then
```
Comment (hassanfa) [inline scripts/novaseq/checkfornewrun.bash:65]: Similar to above. Probably without `-e`.

Comment (barrystokman) [reply scripts/novaseq/checkfornewrun.bash]: fix in https://github.com/Clinical-Genomics/demultiplexing/pull/103/commits/d1099da8290b8899976efb5aaa7e56eddf228fa8

**Comment by barrystokman** [comment]
@hassanfa thanks for the feedback!

**Comment by barrystokman** [comment]
@moahaegglund eveything is implemented, tested and reviewed, can you approve the deployment? I'll take care of actually bumping the version and deployment.

**Comment by moahaegglund** [comment]
Thanks Barry, great work! :)
This week is a bad week for deployment. We will have two S4:s demultiplexing over the weekend and many cases to analyse in MIP next week, which is a short week as well. We can't take the risk of something happening with the demultiplexing now. 
I think more people have to be involved in the decision when to deploy now during summer as bioinfo prod will be even more short-handed now that I leave. @emmser @henrikstranneheim @patrikgrenfeldt

**Comment by patrikgrenfeldt** [comment]
@barrystokman could you have a look and see if you can solve the conflict? I had a look and I think it will perhaps be easier for you

**Comment by barrystokman** [comment]
Predeploy retest:
TC1:
image
Correct sample sheet generated: :heavy_check_mark: 
TC2:
image
Existing sample sheet used: :heavy_check_mark: 
TC3:
image
Empty sample sheet overwritten by correct sample sheet: :heavy_check_mark: 
TC4:
image
Demux not started, sample sheet is empty: :heavy_check_mark:

**Comment by barrystokman** [comment]
deployed on thalamus:
image

**Pull request merged**